domain/machine/state/machine_cloud_instance.go

```diff
@@ -0,0 +1,153 @@
+SELECT (*) AS (&instanceData.*)
+FROM   machine_cloud_instance 
+WHERE  machine_uuid = $instanceData.machine_uuid`
+	machineUUIDQuery := instanceData{MachineUUID: machineUUID}
+	retrieveHardwareCharacteristicsStmt, err := st.Prepare(retrieveHardwareCharacteristics, machineUUIDQuery, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing retrieve hardware characteristics statement")
+	}
+
+	var row instanceData
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, retrieveHardwareCharacteristicsStmt, machineUUIDQuery).Get(&row))
+	}); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, errors.Annotatef(errors.NotFound, "instance data for machine %q", machineUUID)
+		}
+		return nil, errors.Annotatef(domain.CoerceError(err), "querying instance data for machine %q", machineUUID)
+	}
+	return row.toHardwareCharacteristics(), nil
+}
+
+// SetInstanceData sets an entry in the instance data table along with
+// the instance tags and the link to a lxd profile if any.
+func (st *State) SetInstanceData(
+	ctx context.Context,
+	machineUUID string,
+	instanceID instance.Id,
+	hardwareCharacteristics instance.HardwareCharacteristics,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setInstanceData := `
+INSERT INTO machine_cloud_instance (*)
+VALUES ($instanceData.*)
+`
+	setInstanceDataStmt, err := st.Prepare(setInstanceData, instanceData{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setInstanceTags := `
+INSERT INTO instance_tag (*)
+VALUES ($instanceTag.*)
+`
+	setInstanceTagStmt, err := st.Prepare(setInstanceTags, instanceTag{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		instanceData := instanceData{
+			MachineUUID:          machineUUID,
+			InstanceID:           string(instanceID),
+			Arch:                 hardwareCharacteristics.Arch,
+			Mem:                  hardwareCharacteristics.Mem,
+			RootDisk:             hardwareCharacteristics.RootDisk,
+			RootDiskSource:       hardwareCharacteristics.RootDiskSource,
+			CPUCores:             hardwareCharacteristics.CpuCores,
+			CPUPower:             hardwareCharacteristics.CpuPower,
+			AvailabilityZoneUUID: hardwareCharacteristics.AvailabilityZone,
+			VirtType:             hardwareCharacteristics.VirtType,
+		}
+		if err := tx.Query(ctx, setInstanceDataStmt, instanceData).Run(); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "inserting instance data for machine %q", machineUUID)
+		}
+		for _, tag := range *hardwareCharacteristics.Tags {
+			instanceTag := instanceTag{
+				MachineUUID: machineUUID,
+				Tag:         tag,
+			}
+			if err := tx.Query(ctx, setInstanceTagStmt, instanceTag).Run(); err != nil {
+				return errors.Annotatef(domain.CoerceError(err), "inserting instance tag %q for machine %q", tag, machineUUID)
+			}
+		}
```

Comment:
This can be moved to a bulk insert. Sqlair provides this now.

Reply:
Indeed, I wasn't aware of that! Looks amazing, see 29982a633a.